### PR TITLE
fix types: accept a readonly array for configuration options

### PR DESCRIPTION
**Pull request by pcorpet**
The webpack main function can get multiple configuration as an array. However this array doesn't need to be mutable. This fix makes it accept readonly arrays.

**What kind of change does this PR introduce?**

bugfix for typescript types declaration

**Did you add tests for your changes?**

No: I couldn't find typing tests.

**Does this PR introduce a breaking change?**

No: all types that were accepted before are still accepted, it accepts new types.

**What needs to be documented once your changes are merged?**

Nothing, this is a fix.

---
This change is [Reviewable](https://reviewable.io/reviews/webpack/webpack/12818)

### CODE REVIEW

**Comment by sokra** [comment]
This is a generated file. Look into lib/webpack.js or lib/index.js and fix it in the jsdocs comments

**Comment by pcorpet** [comment]
> This is a generated file. Look into lib/webpack.js or lib/index.js and fix it in the jsdocs comments

Thanks @sokra , I hadn't noticed.

I'm now stuck on https://github.com/microsoft/TypeScript/issues/42768

```js
			if (Array.isArray(options)) {
				/** @type {MultiCompiler} */
				compiler = createMultiCompiler(options, options);
				watch = options.some(options => options.watch);
				watchOptions = options.map(options => options.watchOptions || {});
			} else {
				/** @type {Compiler} */
				compiler = createCompiler(options);
				watch = options.watch;
				watchOptions = options.watchOptions || {};
			}
```

The isArray only guards propery against mutable arrays, not readonly arrays (see the bug in Typescript for deeper details).
Maybe we could cast the type of options in the else branch?

**Comment by pcorpet** [comment]
Can someone help me with the failing test please?

File: lib/webpack.js [hunk 1]
```diff
@@ -114,10 +114,12 @@ const webpack = /** @type {WebpackFunctionSingle & WebpackFunctionMulti} */ (
 				watch = options.some(options => options.watch);
 				watchOptions = options.map(options => options.watchOptions || {});
 			} else {
+				/** @type {WebpackOptions} */
+				const webpackOptions = /** @type {WebpackOptions} */ options;
```
Comment (sokra) [inline lib/webpack.js:118]:
```suggestion
				const webpackOptions = /** @type {WebpackOptions} */ (options);
```

**Review by sokra — CHANGES_REQUESTED** [review]

**Review by sokra — APPROVED** [review]

**Pull request closed without merging**